### PR TITLE
Improve CountLeadingZeros() algorithm, nits.

### PR DESCRIPTION
**Pull request by LDj3SNuD**
The (new) proposed solution offers these improvements over the current (old) one:

```
- Linear Test -
	CountLeadingZerosNew_8bit: CS = 6528000; CNT = 6553600
	CountLeadingZerosOld_8bit: CS = 6528000; CNT = 6553600
		tNew - tOld (%) = -10%
	CountLeadingSignsNew_8bit: CS = 6502400; CNT = 6553600
	CountLeadingSignsOld_8bit: CS = 6502400; CNT = 6553600
		tNew - tOld (%) = -24%

- Random Test -
	CountLeadingZerosNew_8bit: CS = 6528563; CNT = 6553600
	CountLeadingZerosOld_8bit: CS = 6529830; CNT = 6553600
		tNew - tOld (%) = -2%
	CountLeadingSignsNew_8bit: CS = 6499957; CNT = 6553600
	CountLeadingSignsOld_8bit: CS = 6501543; CNT = 6553600
		tNew - tOld (%) = -4%

- Linear Test -
	CountLeadingZerosNew_16bit: CS = 6553500; CNT = 6553600
	CountLeadingZerosOld_16bit: CS = 6553500; CNT = 6553600
		tNew - tOld (%) = -14%
	CountLeadingSignsNew_16bit: CS = 6553400; CNT = 6553600
	CountLeadingSignsOld_16bit: CS = 6553400; CNT = 6553600
		tNew - tOld (%) = -15%

- Random Test -
	CountLeadingZerosNew_16bit: CS = 6553101; CNT = 6553600
	CountLeadingZerosOld_16bit: CS = 6547784; CNT = 6553600
		tNew - tOld (%) = -10%
	CountLeadingSignsNew_16bit: CS = 6553838; CNT = 6553600
	CountLeadingSignsOld_16bit: CS = 6546733; CNT = 6553600
		tNew - tOld (%) = -2%

- Linear Test -
	CountLeadingZerosNew_32bit: CS = 16843037; CNT = 16843010
	CountLeadingZerosOld_32bit: CS = 16843037; CNT = 16843010
		tNew - tOld (%) = -5%
	CountLeadingSignsNew_32bit: CS = 16843064; CNT = 16843010
	CountLeadingSignsOld_32bit: CS = 16843064; CNT = 16843010
		tNew - tOld (%) = -8%

- Random Test -
	CountLeadingZerosNew_32bit: CS = 16838946; CNT = 16843010
	CountLeadingZerosOld_32bit: CS = 16847447; CNT = 16843010
		tNew - tOld (%) = -7%
	CountLeadingSignsNew_32bit: CS = 16838791; CNT = 16843010
	CountLeadingSignsOld_32bit: CS = 16848550; CNT = 16843010
		tNew - tOld (%) = -3%

- Linear Test -
	CountLeadingZerosNew_64bit: CS = 16843069; CNT = 16843009
	CountLeadingZerosOld_64bit: CS = 16843069; CNT = 16843009
		tNew - tOld (%) = -5%
	CountLeadingSignsNew_64bit: CS = 16843065; CNT = 16843009
	CountLeadingSignsOld_64bit: CS = 16843065; CNT = 16843009
		tNew - tOld (%) = -12%

- Random Test -
	CountLeadingZerosNew_64bit: CS = 16848358; CNT = 16843009
	CountLeadingZerosOld_64bit: CS = 16851350; CNT = 16843009
		tNew - tOld (%) = -9%
	CountLeadingSignsNew_64bit: CS = 16844479; CNT = 16843009
	CountLeadingSignsOld_64bit: CS = 16854444; CNT = 16843009
		tNew - tOld (%) = -3%
```

### CODE REVIEW

File: ChocolArm64/Instruction/ASoftFallback.cs [hunk 1]
```diff
@@ -38,6 +37,18 @@ public static uint CountSetBits8(uint Value)
             return (Value >> 4) + (Value & 0x0f);
         }
 
+        private static ulong CountSetBits64(ulong Value, int Size)
+        {
+            Value = ((Value >> 1 ) & 0x5555555555555555) + (Value & 0x5555555555555555);
```
Comment (gdkchan) [inline ChocolArm64/Instruction/ASoftFallback.cs]: nit: Why the space after the number (1, and 2/4 on the lines below)?

**Comment by LDj3SNuD** [comment]
The proposed solution, to the dynamic analysis, does not seem to provide better performance than the pre-existing one, in spite of the different class of complexity of the time.

**Comment by gdkchan** [comment]
I made a quick test yesterday. It was faster when testing with random numbers. When using numbers where there is a bit set on one of the higher bits (like 0x7FFFFFFF/0xFFFFFFFF so the loop exits early) the old approach was faster. The old approach had linear search time while the new one has constant time but is slower on the aforementioned cases.

I think something like this should be faster (and simpler) than the current algorithm:
```
private static int CountLeadingZeros32(uint Value)
{
    if (Value == 0)
    {
        return 32;
    }

    int Count = 0;

    while (((Value << Count) & 0xf0000000) == 0)
    {
        Count += 4;
    }

    while (((Value << Count) & 0x80000000u) == 0)
    {
        Count++;
    }

    return Count;
}
```
Not tested through.

**Comment by LDj3SNuD** [comment]
```
public static ulong CountLeadingSigns(ulong Value, int Size)
{
    Value ^= Value >> 1;
    Value &= 0x7fffffffffffffffUL >> (64 - Size);

    return CountLeadingZeros(Value, Size) - 1;
}

public static ulong CountLeadingZeros(ulong Value, int Size)
{
    if (Value == 0)
    {
        return (ulong)Size;
    }

    int Count = 0;

    while (((Value << Count) & (0b1111UL << (Size - 4))) == 0)
    {
        Count += 4;
    }

    while (((Value << Count) & (0b1UL << (Size - 1))) == 0)
    {
        Count++;
    }

    return (ulong)Count;
}
```

This is a fully functional solution, which takes up your idea, adapted to work with the 4 types of size; there are no improvements over the existing solution.

**Comment by LDj3SNuD** [comment]
The (new) proposed solution offers these improvements over the current (old) one:

```
- Linear Test -
	CountLeadingZerosNew_8bit: CS = 6528000; CNT = 6553600
	CountLeadingZerosOld_8bit: CS = 6528000; CNT = 6553600
		tNew - tOld (%) = -10%
	CountLeadingSignsNew_8bit: CS = 6502400; CNT = 6553600
	CountLeadingSignsOld_8bit: CS = 6502400; CNT = 6553600
		tNew - tOld (%) = -24%

- Random Test -
	CountLeadingZerosNew_8bit: CS = 6528563; CNT = 6553600
	CountLeadingZerosOld_8bit: CS = 6529830; CNT = 6553600
		tNew - tOld (%) = -2%
	CountLeadingSignsNew_8bit: CS = 6499957; CNT = 6553600
	CountLeadingSignsOld_8bit: CS = 6501543; CNT = 6553600
		tNew - tOld (%) = -4%

- Linear Test -
	CountLeadingZerosNew_16bit: CS = 6553500; CNT = 6553600
	CountLeadingZerosOld_16bit: CS = 6553500; CNT = 6553600
		tNew - tOld (%) = -14%
	CountLeadingSignsNew_16bit: CS = 6553400; CNT = 6553600
	CountLeadingSignsOld_16bit: CS = 6553400; CNT = 6553600
		tNew - tOld (%) = -15%

- Random Test -
	CountLeadingZerosNew_16bit: CS = 6553101; CNT = 6553600
	CountLeadingZerosOld_16bit: CS = 6547784; CNT = 6553600
		tNew - tOld (%) = -10%
	CountLeadingSignsNew_16bit: CS = 6553838; CNT = 6553600
	CountLeadingSignsOld_16bit: CS = 6546733; CNT = 6553600
		tNew - tOld (%) = -2%

- Linear Test -
	CountLeadingZerosNew_32bit: CS = 16843037; CNT = 16843010
	CountLeadingZerosOld_32bit: CS = 16843037; CNT = 16843010
		tNew - tOld (%) = -5%
	CountLeadingSignsNew_32bit: CS = 16843064; CNT = 16843010
	CountLeadingSignsOld_32bit: CS = 16843064; CNT = 16843010
		tNew - tOld (%) = -8%

- Random Test -
	CountLeadingZerosNew_32bit: CS = 16838946; CNT = 16843010
	CountLeadingZerosOld_32bit: CS = 16847447; CNT = 16843010
		tNew - tOld (%) = -7%
	CountLeadingSignsNew_32bit: CS = 16838791; CNT = 16843010
	CountLeadingSignsOld_32bit: CS = 16848550; CNT = 16843010
		tNew - tOld (%) = -3%

- Linear Test -
	CountLeadingZerosNew_64bit: CS = 16843069; CNT = 16843009
	CountLeadingZerosOld_64bit: CS = 16843069; CNT = 16843009
		tNew - tOld (%) = -5%
	CountLeadingSignsNew_64bit: CS = 16843065; CNT = 16843009
	CountLeadingSignsOld_64bit: CS = 16843065; CNT = 16843009
		tNew - tOld (%) = -12%

- Random Test -
	CountLeadingZerosNew_64bit: CS = 16848358; CNT = 16843009
	CountLeadingZerosOld_64bit: CS = 16851350; CNT = 16843009
		tNew - tOld (%) = -9%
	CountLeadingSignsNew_64bit: CS = 16844479; CNT = 16843009
	CountLeadingSignsOld_64bit: CS = 16854444; CNT = 16843009
		tNew - tOld (%) = -3%
```

**Comment by gdkchan** [comment]
Please fix conflicts.

**Comment by LDj3SNuD** [comment]
Done.

**Comment by gdkchan** [comment]
thanks!

**Pull request merged**